### ISSUE TITLE
munki does offer os updates even if ManagedDeferredInstallDelay is set in com.apple.SoftwareUpdate

### ISSUE THREAD

**Issue by aschwanb** [munki/munki#928]
# Issue
Munki will try to install software updates that are deferred using the`ManagedDeferredInstallDelay` key in the `com.apple.SoftwareUpdate` preference domain.

The install of that software update will fail, leading munki to prompt for the update again after the next run.

# Munki version
```bash
$ managedsoftwareupdate --version
3.6.1.3756
```

# Steps to reproduce
* Use a machine that is missing updates (A macOS 10.14.3 VM missing the 10.14.4 update in this example.)
* Install the following configuration profile
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>PayloadContent</key>
	<array>
		<dict>
			<key>AutomaticCheckEnabled</key>
			<true/>
			<key>AutomaticDownload</key>
			<true/>
			<key>ConfigDataInstall</key>
			<true/>
			<key>ManagedDeferredInstallDelay</key>
			<integer>30</integer>
			<key>PayloadDisplayName</key>
			<string>Software Update</string>
			<key>PayloadIdentifier</key>
			<string>com.github.erikberglund.ProfileCreator.D0CB2A5A-655B-40E6-9E82-A4E8169FB447.com.apple.SoftwareUpdate.F318E866-AB2E-4B5C-8CF8-BC7D433B5E05</string>
			<key>PayloadOrganization</key>
			<string></string>
			<key>PayloadType</key>
			<string>com.apple.SoftwareUpdate</string>
			<key>PayloadUUID</key>
			<string>F318E866-AB2E-4B5C-8CF8-BC7D433B5E05</string>
			<key>PayloadVersion</key>
			<integer>1</integer>
			<key>forceDelayedSoftwareUpdates</key>
			<true/>
		</dict>
	</array>
	<key>PayloadDisplayName</key>
	<string>Software Update </string>
	<key>PayloadIdentifier</key>
	<string>com.github.erikberglund.ProfileCreator.D0CB2A5A-655B-40E6-9E82-A4E8169FB447</string>
	<key>PayloadOrganization</key>
	<string>ProfileCreator</string>
	<key>PayloadScope</key>
	<string>System</string>
	<key>PayloadType</key>
	<string>Configuration</string>
	<key>PayloadUUID</key>
	<string>D0CB2A5A-655B-40E6-9E82-A4E8169FB447</string>
	<key>PayloadVersion</key>
	<integer>1</integer>
</dict>
</plist>
```
* Run `softwareupdate`
```bash
$ softwareupdate -l
Software Update Tool

Finding available software
No new software available.
```

* Run `managedsoftwareupdate`
```bash
$ sudo managedsoftwareupdate --checkonly
Managed Software Update Tool
Copyright 2010-2019 The Munki Project
https://github.com/munki/munki

Starting...
Checking for available updates...
    Retrieving list of software for this machine...
    Retrieving list of software for this machine...
    Preventing idle sleep
    Getting client resources...
    Allowing idle sleep
Checking Apple Software Update catalog...
Checking for available Apple Software Updates...
    Finding available software...
    No updates are available.
    Caching metadata for product ID 041-49230...
	0..20..40..60..80..100
    Caching package metadata for product ID 041-49230...
    [...]
    The following Apple Software Updates are available to install:
        + macOS 10.14.4 Update-
           *Restart required

Run managedsoftwareupdate --installonly to install the downloaded updates.
Finishing...
Done.
```
* This will lead to Managed Software Center prompting the user to log out and update. However, the update will fail.
* Log output from `/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log`
```
Apr 15 2019 17:38:46 +0200 ### Starting managedsoftwareupdate run: logoutinstall ###
Apr 15 2019 17:38:46 +0200 Starting...
Apr 15 2019 17:38:46 +0200     The following Apple Software Updates are available to install:
Apr 15 2019 17:38:46 +0200         + macOS 10.14.4 Update-
Apr 15 2019 17:38:46 +0200            *Restart required
Apr 15 2019 17:38:46 +0200 Installing available Apple Software Updates...
Apr 15 2019 17:38:46 +0200     Caching CatalogURL https://swscan.apple.com/content/catalogs/others/index-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog
Apr 15 2019 17:38:49 +0200     macOS 10.14.4 Update- : No such update
Apr 15 2019 17:38:49 +0200     No updates are available.
Apr 15 2019 17:38:49 +0200 WARNING: Apple update macOS 10.14.4 Update, 041-49230 may have failed to install. No record of success or failure.
Apr 15 2019 17:38:49 +0200 Finishing...
Apr 15 2019 17:38:49 +0200 Saving application inventory...
Apr 15 2019 17:38:54 +0200 ### Ending managedsoftwareupdate run ###
Apr 15 2019 17:38:54 +0200 Software installed or removed requires a restart.
Apr 15 2019 17:38:59 +0200     Restarting now.
Apr 15 2019 17:39:00 +0200     Restarting now.
```
* In our example, the machine is still running macOS 10.14.3
* When `managedsoftwareupdate` is run the next time, it will prompt for the software update again.

# Cause
This is only speculation.

It looks like `softwareupdate` will still populate `/Library/Preferences/com.apple.SoftwareUpdate.plist` with information about missing updates. As far as I could tell, the content of this plist does not change whether the SoftwareUpdate profile is installed or not.
```bash
$ defaults read /Library/Preferences/com.apple.SoftwareUpdate.plist
{
    LastAttemptBuildVersion = "10.14.3 (18D109)";
    LastAttemptSystemVersion = "10.14.3 (18D109)";
    LastBackgroundSuccessfulDate = "2019-03-13 11:21:27 +0000";
    LastFullSuccessfulDate = "2019-04-15 15:54:41 +0000";
    LastRecommendedUpdatesAvailable = 1;
    LastResultCode = 0;
    LastSuccessfulDate = "2019-04-15 15:54:41 +0000";
    LastUpdatesAvailable = 1;
    PrimaryLanguages =     (
        en,
        "en-GB"
    );
    RecommendedUpdates =     (
                {
            "Display Name" = "macOS 10.14.4 Update";
            "Display Version" = " ";
            Identifier = "macOS 10.14.4 Update";
            "Product Key" = "041-49230";
        }
    );
    SkipLocalCDN = 0;
}
```
I think munki is searching this plist to find updates it needs to install. This would explain why the update is offered through Managed Software Center even though `softwareupdate -l` does not show it. If this is the case, the issue could be addressed by checking if the content of `/Library/Preferences/com.apple.SoftwareUpdate.plist` matches the output of `softwareupdate -l`.

**Comment by gregneagle** [comment]
I'd encourage you to build from the Munki3dev branch and try that version.

**Comment by gregneagle** [comment]
See also https://groups.google.com/d/topic/munki-dev/CFeP-zbtBXY/discussion

**Comment by aschwanb** [comment]
Thanks. I'll get back to you as soon as I'm done testing.

**Comment by aschwanb** [comment]
I did build from Munki3dev and can confirm that the software updates offered by munki are now consistent with the output of `softwareupdate -l`. This fixes the issue.

```bash
$ managedsoftwareupdate --version
3.6.2.3773
```

**Comment by gregneagle** [comment]
OK, thanks. I'll close this issue as being addressed by the upcoming 3.6.2 release.
